server/src/main/java/org/elasticsearch/lucene/search/uhighlight/CustomUnifiedHighlighter.java

```diff
@@ -272,4 +229,63 @@ protected OffsetSource getOffsetSource(String field) {
         }
         return offsetSource;
     }
+
+    /**
+     * Returns true if the provided {@link Query} is not compatible with the {@link HighlightFlag#WEIGHT_MATCHES}
+     * mode of this highlighter.
+     *
+     * @param query The query to highlight
+     */
+    private boolean weightMatchesUnsupported(Query query) {
+        boolean[] hasUnknownLeaf = new boolean[1];
+        query.visit(new QueryVisitor() {
+            @Override
+            public void visitLeaf(Query leafQuery) {
+                /**
+                 * The parent-child query requires to load global ordinals and to access
+                 * documents outside of the scope of the highlighted doc.
+                 * We disable the {@link HighlightFlag#WEIGHT_MATCHES} mode in this case
+                 * in order to preserve the compatibility.
+                 */
+                if (leafQuery.getClass().getSimpleName().equals("LateParsingQuery")) {
+                    hasUnknownLeaf[0] = true;
+                }
+                super.visitLeaf(query);
+            }
+
+            @Override
+            public void consumeTerms(Query leafQuery, Term... terms) {
+                if (leafQuery instanceof AbstractScriptFieldQuery) {
+                    /**
+                     * Queries on runtime fields don't support the matches API.
+                     */
```

Comment:
Maybe add a TODO here - it would be nice to get this working for keyword script fields at the least.

Reply:
Agreed, I pushed https://github.com/elastic/elasticsearch/pull/96068/commits/9cd6e3ddcec77a8ea0fbf6c3c08efb46e9b0832d